File: frontend/src/services/walletService.ts
import api from './api';

export interface Wallet {
  id: string;
  customerId?: string;
  vendorId?: string;
  balance: number;
  createdAt: string;
  updatedAt: string;
}

export interface WalletTransaction {
  id: string;
  walletId: string;
  amount: number;
  type: 'cash_in' | 'payment' | 'refund' | 'vendor_payout' | 'withdrawal';
  referenceId?: string;
  status: 'pending' | 'success' | 'failed';
  paymentMethod?: string;
  createdAt: string;
}

export interface VendorWithdrawal {
  id: string;
  vendorId: string;
  amount: number;
  status: 'pending' | 'processing' | 'completed' | 'failed';
  accountName: string;
  accountNumber: string;
  accountType: 'bank' | 'ewallet';
  accountProvider: string;
  notes?: string;
  createdAt: string;
  updatedAt: string;
}

export async function getWallet(): Promise<Wallet> {
  const response = await api.get<Wallet>('/wallet/balance');
  return response.data;
}

export interface PaginatedTransactions {
  data: WalletTransaction[];
  total: number;
  page: number;
  limit: number;
}

export async function getTransactions(page = 1, limit = 15): Promise<PaginatedTransactions> {
  const response = await api.get<PaginatedTransactions>('/wallet/transactions', {
    params: { page, limit },
  });
  return response.data;
}

export async function initiateCashIn(
  amount: number,
  paymentMethod: 'GCASH' | 'MAYA' | 'QRPH',
): Promise<{ transactionId: string; checkoutUrl: string; status: string }> {
  const response = await api.post<{ transactionId: string; checkoutUrl: string; status: string }>(
    '/wallet/cash-in',
    { amount, paymentMethod },
  );
  return response.data;
}

export async function getVendorWallet(): Promise<Wallet | { needsSetup: boolean }> {
  const response = await api.get<Wallet>('/wallet/vendor');
  return response.data;
}

export async function getVendorTransactions(): Promise<WalletTransaction[]> {
  const response = await api.get<WalletTransaction[]>('/wallet/vendor/transactions');
  return response.data;
}

export async function requestVendorWithdrawal(
  amount: number,
  accountDetails: {
    accountName: string;
    accountNumber: string;
    accountType: 'bank' | 'ewallet';
    accountProvider: string;
  },
): Promise<VendorWithdrawal> {
  const response = await api.post<VendorWithdrawal>('/wallet/vendor/withdraw', {
    amount,
    ...accountDetails,
  });
  return response.data;
}

export async function getVendorWithdrawals(): Promise<VendorWithdrawal[]> {
  const response = await api.get<VendorWithdrawal[]>('/wallet/vendor/withdrawals');
  return response.data;
}

File: flutter_geobite/lib/services/wallet_service.dart
import '../core/api_client.dart';

class WalletService {
  /// Get the current user's wallet balance
  Future<Map<String, dynamic>> getWallet() async {
    try {
      final response = await apiClient.dio.get('/wallet/balance');
      return Map<String, dynamic>.from(response.data);
    } catch (e) {
      throw Exception('Failed to load wallet: $e');
    }
  }

  /// Get paginated wallet transactions
  Future<Map<String, dynamic>> getTransactions({int page = 1, int limit = 15}) async {
    try {
      final response = await apiClient.dio.get(
        '/wallet/transactions',
        queryParameters: {'page': page, 'limit': limit},
      );
      return Map<String, dynamic>.from(response.data);
    } catch (e) {
      throw Exception('Failed to load transactions: $e');
    }
  }

  /// Initiate a cash-in / top-up
  Future<Map<String, dynamic>> initiateCashIn(double amount, String paymentMethod) async {
    try {
      final response = await apiClient.dio.post('/wallet/cash-in', data: {
        'amount': amount,
        'paymentMethod': paymentMethod,
      });
      return Map<String, dynamic>.from(response.data);
    } catch (e) {
      throw Exception('Failed to initiate cash-in: $e');
    }
  }

  // ── Vendor wallet ──────────────────────────────────────────────────────────

  Future<Map<String, dynamic>> initiateVendorCashIn(double amount, String paymentMethod) async {
    try {
      final response = await apiClient.dio.post('/wallet/vendor/cash-in', data: {
        'amount': amount,
        'paymentMethod': paymentMethod,
      });
      return Map<String, dynamic>.from(response.data);
    } catch (e) {
      throw Exception('Failed to initiate vendor cash-in: $e');
    }
  }

  Future<Map<String, dynamic>> getVendorWallet() async {
    try {
      final response = await apiClient.dio.get('/wallet/vendor');
      return Map<String, dynamic>.from(response.data);
    } catch (e) {
      throw Exception('Failed to load vendor wallet: $e');
    }
  }

  Future<List<Map<String, dynamic>>> getVendorTransactions() async {
    try {
      final response = await apiClient.dio.get('/wallet/vendor/transactions');
      final List list = response.data is List ? response.data : [];
      return list.cast<Map<String, dynamic>>();
    } catch (e) {
      throw Exception('Failed to load vendor transactions: $e');
    }
  }

  Future<Map<String, dynamic>> requestVendorWithdrawal(
    double amount,
    Map<String, String> accountDetails,
  ) async {
    try {
      final response = await apiClient.dio.post('/wallet/vendor/withdraw', data: {
        'amount': amount,
        ...accountDetails,
      });
      return Map<String, dynamic>.from(response.data);
    } catch (e) {
      throw Exception('Failed to request vendor withdrawal: $e');
    }
  }

  Future<List<Map<String, dynamic>>> getVendorWithdrawals() async {
    try {
      final response = await apiClient.dio.get('/wallet/vendor/withdrawals');
      final List list = response.data is List ? response.data : [];
      return list.cast<Map<String, dynamic>>();
    } catch (e) {
      throw Exception('Failed to load vendor withdrawals: $e');
    }
  }

  // ── Customer withdrawals ───────────────────────────────────────────────────

  Future<Map<String, dynamic>> requestCustomerWithdrawal(
    double amount,
    Map<String, String> accountDetails,
  ) async {
    try {
      final response = await apiClient.dio.post('/wallet/withdraw', data: {
        'amount': amount,
        ...accountDetails,
      });
      return Map<String, dynamic>.from(response.data);
    } catch (e) {
      throw Exception('Failed to request withdrawal: $e');
    }
  }

  Future<List<Map<String, dynamic>>> getCustomerWithdrawals() async {
    try {
      final response = await apiClient.dio.get('/wallet/withdrawals');
      final List list = response.data is List ? response.data : [];
      return list.cast<Map<String, dynamic>>();
    } catch (e) {
      throw Exception('Failed to load withdrawals: $e');
    }
  }
  /// Get rider earnings grouped by date, plus weekly total
  Future<Map<String, dynamic>> getRiderEarnings() async {
    try {
      final response = await apiClient.dio.get('/wallet/rider/earnings');
      return Map<String, dynamic>.from(response.data);
    } catch (e) {
      throw Exception('Failed to load rider earnings: $e');
    }
  }
}

final walletService = WalletService();
